feat: leave type wise leave block dates

rtdany10 opened the pull request:
This PR gives user the ability to block leave days by leave type.

https://user-images.githubusercontent.com/52111700/198834333-07bf7d6e-f242-4998-b978-1ea317b4bded.mp4

rtdany10 committed:
fix: specify doctype in query

rtdany10 committed:
Merge branch 'develop' of https://github.com/frappe/hrms into develop

rtdany10 committed:
feat: leave type in leave block

rtdany10 committed:
feat: bulk add days

rtdany10 committed:
chore: make form dirty

rtdany10 committed:
chore: pre-commit

rtdany10 commented:
Unrelated test failing.
@ruchamahabal please review the PR when you get time

rtdany10 commented:
@ruchamahabal :grimacing:

rtdany10 commented:
@ruchamahabal :grimacing:

rtdany10 committed:
Merge branch 'frappe:develop' into develop

rtdany10 commented:
@ruchamahabal :grimacing:

ruchamahabal commented on hrms/hr/doctype/leave_block_list/leave_block_list.js:
```diff
@@ -3,6 +3,101 @@
+						fieldname: 'thursday',
+						fieldtype: 'Check'
+					},
+					{
+						label: 'Sunday',
+						fieldname: 'sunday',
+						fieldtype: 'Check'
+					},
+					{
+						fieldname: 'col_break_0',
+						fieldtype: 'Column Break'
+					},
+					{
+						label: 'Tuesday',
+						fieldname: 'tuesday',
+						fieldtype: 'Check'
+					},
+					{
+						label: 'Friday',
+						fieldname: 'friday',
+						fieldtype: 'Check'
+					},
+					{
+						fieldname: 'col_break_0',
+						fieldtype: 'Column Break'
+					},
+					{
+						label: 'Wednesday',
+						fieldname: 'wednesday',
+						fieldtype: 'Check'
+					},
+					{
+						label: 'Saturday',
+						fieldname: 'saturday',
+						fieldtype: 'Check'
+					},
+					{
+						fieldname: 'sec_break_0',
+						fieldtype: 'Section Break'
+					},
```
Try using **Multicheck** control instead?
Docs: https://frappeframework.com/docs/v14/user/en/api/controls

rtdany10 replied:
Done, we also need to update the documentation of **MultiCheck**.

ruchamahabal commented on hrms/hr/doctype/leave_block_list/leave_block_list.py:
```diff
@@ -7,37 +7,56 @@
 			dates.append(d.block_date)
 
+	@frappe.whitelist()
+	def get_weekly_off_dates(self, start_date, end_date, days, reason):
```
Change the name to `set_weekly_off_dates` since it's not returning the dates but adding them to the table instead.

rtdany10 replied:
Changed

ruchamahabal commented on hrms/hr/doctype/leave_block_list/leave_block_list.js:
```diff
@@ -3,6 +3,101 @@
 
 frappe.ui.form.on('Leave Block List', {
 	refresh: function(frm) {
+		frm.add_custom_button(__("Bulk Add Dates"), function () {
```
Add this button as a field above the dates table instead, since this action is associated with adding dates to the **Leave Block List Dates** table. 

Other forms where this pattern is used:
<img width="1295" alt="image" src="https://user-images.githubusercontent.com/24353136/199908174-c68be3a8-b6c8-4525-9e1d-980494705aba.png">
 
<img width="1295" alt="image" src="https://user-images.githubusercontent.com/24353136/199908517-5cf6d73a-db32-4263-9e50-1b607fa262b5.png">

rtdany10 replied:
Done, also changed the label to `Add Day-wise Dates`

ruchamahabal reviewed (changes requested):
Add test cases in the "leave block list" test controller for getting applicable block dates with leave type (single and all lists)

rtdany10 committed:
feat: tests and review changes

rtdany10 committed:
Merge branch 'develop' of https://github.com/rtdany10/hrms into develop

rtdany10 committed:
chore: pre-commit

rtdany10 commented:
@ruchamahabal I have added test cases, also modified the test data; added `applies_to_all_departments` on the first test record. Hope it's okay. We should add department wise test as well.

rtdany10 commented:
@ruchamahabal please check

ruchamahabal commented on hrms/hr/doctype/leave_block_list/leave_block_list.py:
```diff
@@ -49,30 +68,30 @@ def get_applicable_block_lists(employee=None, company=None, all_lists=False):
-	):
-		add_block_list(block_list)
+	conditions = {"applies_to_all_departments": 1, "company": company}
+	if leave_type:
+		conditions["leave_type"] = leave_type
```
You will have to add `or_filters` here to check:
- if the leave type matches
- or the leave type is not set in the list (meaning it's applicable for all leave types)

Currently, it does not work in the second case where the block list is not set up leave-type-wise.

rtdany10 replied:
Thank you, have fixed the same :)

ruchamahabal commented:
> added `applies_to_all_departments` on the first test record. Hope it's okay. 

Any reason to do this?

rtdany10 committed:
fix: consider global block dates for leave type

rtdany10 commented:
> > added `applies_to_all_departments` on the first test record. Hope it's okay.
> 
> Any reason to do this?

None of the test actually check for it, so while adding new test records, I added that key as well :grimacing:

rtdany10 commented:
@ruchamahabal :grimacing:

rtdany10 commented:
@ruchamahabal :grimacing:

ruchamahabal commented:
@rtdany10 thank you for the contribution. 

A note for the future contributions: Avoid commit messages like `feat: tests and review changes`
Use messages describing the actual fix/feature/chore being implemented.

rtdany10 commented:
Noted, will keep the commit message descriptive from now on.
Thank you @ruchamahabal
